testcases/crypto/abfunc.c

```diff
@@ -0,0 +1,1079 @@
+                                    pubkey, privkey);
+}
+
+// Expected to fail
+CK_RV generateABDilithiumKey(CK_SESSION_HANDLE session,
+                             CK_OBJECT_HANDLE *pubkey, CK_OBJECT_HANDLE *privkey)
+{
+    CK_MECHANISM mech = { CKM_IBM_DILITHIUM, NULL, 0 };
+    CK_BBOOL cktrue = TRUE;
+    CK_ATTRIBUTE pubkeyTemplate[] = {
+        {CKA_VERIFY,        &cktrue, sizeof(cktrue)},
+        {CKA_IBM_ATTRBOUND, &cktrue, sizeof(cktrue)}
+    };
+    CK_ATTRIBUTE privkeyTemplate[] = {
+        {CKA_SIGN,          &cktrue, sizeof(cktrue)},
+        {CKA_IBM_ATTRBOUND, &cktrue, sizeof(cktrue)},
+        {CKA_SENSITIVE,     &cktrue, sizeof(cktrue)}
+    };
+
+    return funcs->C_GenerateKeyPair(session, &mech,
+                                    pubkeyTemplate, ARRAY_SIZE(pubkeyTemplate),
+                                    privkeyTemplate, ARRAY_SIZE(privkeyTemplate),
+                                    pubkey, privkey);
+}
+
+void teardown(CK_SESSION_HANDLE session) {
+    int i;
+
+    for (i = 0; i < NUMKEYS; ++i)
+        funcs->C_DestroyObject(session, keys.keys[i]);
+}
+
+/// Test functions
+
+CK_RV do_CheckMechanismInfo()
+{
+    CK_IBM_ATTRIBUTEBOUND_WRAP_PARAMS params = {
+        .hSignVerifyKey = CK_INVALID_HANDLE
+    };
+    CK_MECHANISM mech = {
+        .mechanism = CKM_IBM_ATTRIBUTEBOUND_WRAP,
+        .pParameter = &params,
+        .ulParameterLen = sizeof(params)
+    };
+    CK_RV res = CKR_OK;
+
+    // begin testsuite
+    testsuite_begin("Check CKM_IBM_ATTRIBUTEBOUND_WRAP mechanism info.");
+
+    // skip tests if the slot doesn't support this mechanism
+    if (!mech_supported(SLOT_ID, CKM_IBM_ATTRIBUTEBOUND_WRAP)) {
+        testsuite_skip(2,
+                      "Slot %u doesn't support CKM_IBM_ATTRIBUTEBOUND_WRAP (%u)",
+                      (unsigned int) SLOT_ID,
+                      "CKM_IBM_ATTRIBUTEBOUND_WRAP",
+                      (unsigned int) CKM_IBM_ATTRIBUTEBOUND_WRAP);
+        res = CKR_FUNCTION_NOT_SUPPORTED;
+        goto testcase_cleanup;
+    }
+    testcase_new_assertion();
+    testcase_begin("CKM_IBM_ATTRIBUTEBOUND_WRAP supported for wrapping");
+    if (!wrap_supported(SLOT_ID, mech))
+        testcase_fail("CKM_IBM_ATTRIBUTEBOUND_WRAP does NOT support wrapping");
+    else
+        testcase_pass("CKM_IBM_ATTRIBUTEBOUND_WRAP does support wrapping");
+    testcase_new_assertion();
+    testcase_begin("CKM_IBM_ATTRIBUTEBOUND_WRAP supported for unwrapping");
+    if (!unwrap_supported(SLOT_ID, mech))
+        testcase_fail("CKM_IBM_ATTRIBUTEBOUND_WRAP does NOT support unwrapping");
+    else
+        testcase_pass("CKM_IBM_ATTRIBUTEBOUND_WRAP does support unwrapping");    
+ testcase_cleanup:
+    return res;
+}
+
+void do_TestInvalidKeys(CK_SESSION_HANDLE session)
+{
+    CK_OBJECT_HANDLE pubkey = CK_INVALID_HANDLE, privkey = CK_INVALID_HANDLE;
+    CK_RV rc;
+
+    // Expected to fail with CKR_TEMPLATE_INCOMPLETE
+    testcase_begin("Create dilithium AB key");
```

Comment:
Are you using Dilithium keys here for a specific reason? Dilithium keys are only supported by quite recent hardware. So this tests will be skipped if run on a system with older hardware.  Wouldn't it be more useful to choose a key type that is supported on all hardware, e.g. EC keys?

Reply:
Well, that test is the only test I found for a key type that is not supported for AB keys.  EC is supported, so I cannot use this.  I did not find any other key type supported by EP11 that does not support AB...